### PR TITLE
feat: Save objects states in Obj Manager to keep them btwn rooms

### PR DESCRIPTION
**Pull request by StraToN**
Fixes https://github.com/godot-escoria/escoria-issues/issues/193

Unblocks https://github.com/godot-escoria/escoria-demo-game/pull/532

### CODE REVIEW

**Comment by BHSDuncan** [comment]
How are we going to merge this? If we put it into `develop` before we merge the "instant transition" PR, it's going to cause merge conflicts, especially given how very much different the object and room managers have become.

How do we want to handle this?

**Review by dploeger — DISMISSED** [review]
Thanks! 🎁

**Comment by dploeger** [comment]
Sorry, didn't read your comment, @BHSDuncan. Maybe you could incorporate this change in your branch then? It isn't quite big.

**Comment by StraToN** [comment]
@BHSDuncan Agree with @dploeger . I'm fine if you cherry-pick this commit into #515 and we close this one.

**Comment by StraToN** [comment]
Status: waiting for #515 to be merged. Will have to be rebased after that.

**Comment by StraToN** [comment]
To test:
- checkout [branch "room6" from @balloonpopper ](https://github.com/balloonpopper/escoria-demo-game/tree/room6)
- rebase this branch on origin/develop
- If you want to see the issue, test (apply scenario below)
- apply this PR onto the resulting branch
- run

Scenario:
1. Access room06 -> notice that right door is closed
2. exit room06 to room05
3. Get back to room06
-> if PR not applied, notice that right door is now open
-> else, notice that right door is still closed

**Comment by balloonpopper** [comment]
The state was preserved in my testing - the door stays "closed" with the fix applied.
Other issues were found in the transition between these rooms (as dicsussed in the channel) but they seem unrelated to this PR.

File: addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd
Comment (BHSDuncan) [inline addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd:220]: In the future, wouldn't this prevent us from being able to load multiple rooms (for whatever reason, e.g. pre-loading) that contain the same object at the same time? Would it make sense to include these states as part of the larger "room->object" structure for the object manager? Or is this just not a concern?

**Comment by BHSDuncan** [comment]
> The state was preserved in my testing - the door stays "closed" with the fix applied. Other issues were found in the transition between these rooms (as dicsussed in the channel) but they seem unrelated to this PR.

Can you also maybe check if it's preserved across saving and loading? It should be based on what I see, but best to test. :)

**Review by balloonpopper — APPROVED** [review]

**Comment by balloonpopper** [comment]
> Can you also maybe check if it's preserved across saving and loading? It should be based on what I see, but best to test. :)

I was asked to hold off on save testing while @StraToN does some more work on the save system.

**Review by balloonpopper — APPROVED** [review]

File: addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd
Comment (BHSDuncan) [inline addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd:169]:
```suggestion
	# Object exists in room, set it to is last state (if different from "default")
```

File: addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd
Comment (BHSDuncan) [inline addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd:216]: There needs to be a null check based on what `get_object` returns, since returning null is a possibility.

File: addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd
Comment (BHSDuncan) [inline addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd:450]: I'm curious about this change. I remember making this method to check this very specific combination in case we ever have two rooms setup but that are different instances. What benefit do we gain by removing that part of the check? I don't think it does any harm to leave it as it was.

**Review by BHSDuncan — CHANGES_REQUESTED** [review]

File: addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd
Comment (StraToN) [reply addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd:463]: This is because when I save the ESCItem state in the Dictionary, the key is composed of `global_id` and `room_instance_id`. However, when I exit the room to enter another one, then go back to the first room, the `room_instance_id` saved as key is different from the new `room_instance_id` of the room I just went back.

I can revert this change, but then I'll have to create a new test function that doesn't test the instance id.

Comment (BHSDuncan) [reply addons/escoria-core/game/core-scripts/esc/esc_object_manager.gd]: In other words, since the room_object container is no longer deleted when you leave a room if an object's state has changed, room equality is now an issue. Got it. Will take a look.

**Review by BHSDuncan — APPROVED** [review]

**Pull request merged**
